src/coreclr/jit/codegenxarch.cpp

```diff
@@ -9091,13 +9091,176 @@ void CodeGen::genProfilingLeaveCallback(unsigned helper)
 
 #endif // PROFILING_SUPPORTED
 
+#ifdef TARGET_AMD64
+
+//------------------------------------------------------------------------
+// genOSRRecordTier0CalleeSavedRegistersAndFrame: for OSR methods, record the
+//  subset of callee saves already saved by the Tier0 method, and the frame
+//  created by Tier0.
+//
+void CodeGen::genOSRRecordTier0CalleeSavedRegistersAndFrame()
+{
+    assert(compiler->compGeneratingProlog);
+    assert(compiler->opts.IsOSR());
+    assert(compiler->funCurrentFunc()->funKind == FuncKind::FUNC_ROOT);
+
+#if ETW_EBP_FRAMED
+    if (!isFramePointerUsed() && regSet.rsRegsModified(RBM_FPBASE))
+    {
+        noway_assert(!"Used register RBM_FPBASE as a scratch register!");
+    }
+#endif
+
+    // Figure out which set of int callee saves was already saved by Tier0.
+    // Emit appropriate unwind.
+    //
+    PatchpointInfo* const patchpointInfo             = compiler->info.compPatchpointInfo;
+    regMaskTP const       tier0CalleeSaves           = (regMaskTP)patchpointInfo->CalleeSaveRegisters();
+    regMaskTP             tier0IntCalleeSaves        = tier0CalleeSaves & RBM_OSR_INT_CALLEE_SAVED;
+    int const             tier0IntCalleeSaveUsedSize = genCountBits(tier0IntCalleeSaves) * REGSIZE_BYTES;
+
+    JITDUMP("--OSR--- tier0 has already saved ");
+    JITDUMPEXEC(dspRegMask(tier0IntCalleeSaves));
+    JITDUMP("\n");
+
+    // We must account for the Tier0 callee saves.
+    //
+    // These have already happened at method entry; all these
+    // unwind records should be at offset 0.
+    //
+    // RBP is always aved by Tier0 and always pushed first.
+    //
+    assert((tier0IntCalleeSaves & RBM_FPBASE) == RBM_FPBASE);
+    compiler->unwindPush(REG_RBP);
+    tier0IntCalleeSaves &= ~RBM_FPBASE;
+
+    // Now the rest of the Tier0 callee saves.
+    //
+    for (regNumber reg = REG_INT_LAST; tier0IntCalleeSaves != RBM_NONE; reg = REG_PREV(reg))
+    {
+        regMaskTP regBit = genRegMask(reg);
+
+        if ((regBit & tier0IntCalleeSaves) != 0)
+        {
+            compiler->unwindPush(reg);
+        }
+        tier0IntCalleeSaves &= ~regBit;
+    }
```

Comment:
Maybe consider using `genFindLowestReg` here? Took me a bit to understand this loop.

Reply:
This is working backwards through the "preferred pop order" for registers. It needs to match what we do in epilog gen (where we walk from `REG_INT_FIRST` via `REG_NEXT`.

I don't know if that corresponds to numeric order.

I also don't know if the order actually matters for x64 (as long as prolog, epilog, and unwind agree) -- I think the ordering requirement may be an x86-ism where the code manager does a lot of inspection of the jitted codegen rather than relying on unwind. In particular for x64 OSR there can now be two disjoint ranges of pushes/pops; each range will be in the proper order but overall the registers may be handled out of order.

Reply:
Ah, I see. That makes sense.